sdk/src/azure/core/az_mqtt5_rpc_server_hfsm.c

```diff
@@ -0,0 +1,598 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <azure/core/az_mqtt5.h>
+#include <azure/core/az_mqtt5_rpc_server.h>
+#include <azure/core/az_platform.h>
+#include <azure/core/az_result.h>
+#include <azure/core/internal/az_log_internal.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "mqtt_protocol.h"
+
+#include <azure/core/_az_cfg.h>
+
+static az_result root(az_event_policy* me, az_event event);
+static az_result subscribing(az_event_policy* me, az_event event);
+static az_result waiting(az_event_policy* me, az_event event);
+static az_result faulted(az_event_policy* me, az_event event);
+
+AZ_INLINE az_result _handle_request(az_mqtt5_rpc_server* me, az_mqtt5_recv_data* data);
+AZ_NODISCARD az_result _az_rpc_server_policy_init(
+    _az_hfsm* hfsm,
+    _az_event_client* event_client,
+    az_mqtt5_connection* connection);
+
+static az_event_policy_handler _get_parent(az_event_policy_handler child_state)
+{
+  az_event_policy_handler parent_state;
+
+  if (child_state == root)
+  {
+    parent_state = NULL;
+  }
+  else if (child_state == subscribing || child_state == waiting || child_state == faulted)
+  {
+    parent_state = root;
+  }
+  else
+  {
+    // Unknown state.
+    az_platform_critical_error();
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+static az_result root(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  (void)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_rpc_server"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      break;
+
+    case AZ_HFSM_EVENT_ERROR:
+    {
+      if (az_result_failed(az_event_policy_send_inbound_event(me, event)))
+      {
+        az_platform_critical_error();
+      }
+      break;
+    }
+
+    case AZ_HFSM_EVENT_EXIT:
+    {
+      if (_az_LOG_SHOULD_WRITE(AZ_HFSM_EVENT_EXIT))
+      {
+        _az_LOG_WRITE(AZ_HFSM_EVENT_EXIT, AZ_SPAN_FROM_STR("az_mqtt5_rpc_server: PANIC!"));
+      }
+
+      az_platform_critical_error();
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUBACK_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_OPEN_REQ:
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_CLOSE_REQ:
+    case AZ_MQTT5_EVENT_DISCONNECT_RSP:
+      break;
+
+    default:
+      // TODO
```

Comment:
Until we have the event filtering, I'm going to leave this here since an event from a new client would send us to this point unless we explicitly ignore it